libs/grok/src/main/java/org/elasticsearch/grok/ThreadWatchdog.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.grok;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledFuture;
+import java.util.function.BiFunction;
+import java.util.function.LongSupplier;
+
+/**
+ * Protects against long running operations that happen between the register and unregister invocations.
+ * Threads that invoke {@link #register()}, but take too long to invoke the {@link #unregister()} method
+ * will be interrupted.
+ *
+ * This is needed for Joni's {@link org.joni.Matcher#search(int, int, int)} method, because
+ * it can end up spinning endlessly if the regular expression is too complex. Joni has checks
+ * that for every 30k iterations it checks if the current thread is interrupted and if so
+ * returns {@link org.joni.Matcher#INTERRUPTED}.
+ */
+public interface ThreadWatchdog {
+    
+    /**
+     * Registers the current thread and interrupts the current thread
+     * if the takes too long for this thread to invoke {@link #unregister()}.
+     */
+    void register();
+    
+    /**
+     * @return The maximum allowed time for a thread to invoke {@link #unregister()} after {@link #register()}
+     *         has been invoked before this ThreadWatchDog starts to interrupting that thread.
+     */
+    long maxExecutionTime();
```

Comment:
Should this be a `TimeValue`?

Reply:
`TimeValue` is not accessible in the grok module. I don't think we want to make grok module depend on elasticsearch-core module?

Reply:
Ah, that's a shame. I agree. Can we make the name of this method reflect that the time is represented in milliseconds then?

Reply:
yes: c21a40824e033493c191e2f0de1d2474b25b633a